lib/Monads/trace/Trace_Monad.thy

```diff
@@ -68,6 +67,17 @@ text \<open>Returns monad results, ignoring failures and traces.\<close>
 definition mres :: "((tmid \<times> 's) list \<times> ('s, 'a) tmres) set \<Rightarrow> ('a \<times> 's) set" where
   "mres r = Result -` (snd ` r)"
 
+text \<open>True if the monad has a computation resulting in Failed.\<close>
+definition failed :: "((tmid \<times> 's) list \<times> ('s, 'a) tmres) set \<Rightarrow> bool" where
+  "failed r \<equiv> Failed \<in> snd ` r"
```

Comment:
I like it. I'm almost tempted to try it out on the nondet monad as well. But let's resist temptation for now, it'd lead to much more change on that side.